tinygrad/codegen/uops.py

```diff
@@ -386,11 +386,14 @@ def linearize(self, extra_pm:Optional[PatternMatcher]=None, do_type_verify=True)
 
     # fixup gated stores with an IF block to save extra local loads
     @functools.lru_cache(None)
-    def _dfs(u:UOp, gate:UOp) -> UOp:
-      if u.op is UOps.LOAD and u.src[-1].op is UOps.BARRIER:
+    def _dfs(u:UOp, gate:Union[UOp|None]=None, phi:Union[UOp|None]=None) -> UOp:
```

Comment:
is late_gate reset enough for correctness in your tests? Was fine for me https://github.com/tinygrad/tinygrad/pull/5346

We have to write a simpler IF rewrite for multireduce at some point, but let's focus on correctness before speed.

Reply:
reseting the late gate is enough for correctness, this is so that the dfs will propagate through reduction to visit potential preceding reduceops and give them their own if blocks